### ISSUE TITLE
rc: run rc1, rc3 on each broker independently

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#2614]
It seems evident that the current strategy of loading all modules from rank 0 at instance startup isn't going to work on the system instance. I think we'll want to have the rc scripts run on each rank independently, with conditional logic in the scripts based on `flux getattr rank` replacing the current remote execs.

Since modules that are loaded across an instance typically require "upstream" to be initialized before "downstream", we will want to complete module initialization on a given rank before  accepting messages from downstream peers.   (Currently the TBON fully wires up, then modules are loaded so that's a change to the wire-up sequence).

@grondo raised the question in #2612:
>  is the plan to use the same strategy for non-system instances, e.g. flux start -s and instances started by an existing resource manager?

I don't see a compelling reason not to use the same startup scheme in all cases though maybe I'm missing one!

**Comment by grondo** [comment]
> I don't see a compelling reason not to use the same startup scheme in all cases though maybe I'm missing one!

I guess one thing to consider is if a switch to this scheme would impact the startup time of sub-instances, since this would affect every job within the system instance. This design might actually reduce the startup time, but it would be nice to test before deciding to drop the other option.

**Comment by dongahn** [comment]
It might be good to use `flux-tree` to test the performance out.

**Comment by garlick** [comment]
Where does `flux-tree` live?

**Comment by dongahn** [comment]
It is in a branch currently: https://github.com/dongahn/flux-sched/blob/ftree/src/cmd/flux-tree. But this will be my next priority as UQP will need to make use of it.

**Comment by grondo** [comment]
I'm not sure the number of levels of instance hierarchy would affect the startup of each individual instance with this new scheme. I would guess it would be more dependent on the number of levels in the TBON. So for me a scalability study over instance size vs branching factor of the TBON.

**Comment by dongahn** [comment]
@grondo: Good point. I was thinking many brokers trying to access a file system at the same time would be something to watch out for with the new "all-access" scheme vs. the old "broadcast" scheme. If you run `flux-tree` at a large scale, the effect of both the TBON and instance hierarchies can be observed (as a stress case?)

**Comment by grondo** [comment]
> I was thinking many brokers trying to access a file system at the same time would be something to watch out for with the new "all-access" scheme vs. the old "broadcast" scheme. If you run flux-tree at a large scale, the effect of both the TBON and instance hierarchies can be observed (as a stress case?)

Good point!

**Comment by garlick** [comment]
Wanted to jot down these notes (somewhat mid thought):

Proposed new startup sequence for a broker:
1. rank > 0 establish connection with upstream peer
2. once upstream connection is established, run rc1 to load modules locally (rank 0 loads state from disk)
3. begin allowing downstream peers to connect
4. rank 0 starts initial program once configurable "quorum" is connected (full for non-instance, probably)

Shutdown sequence (any broker receiving instance shutdown event, or local systemd stop):
1. tell downstream peers (if any) of impending shutdown, wait for them to disconnect (timeout)
2. once downstream peers disconnect, run rc3 to unload modules locally (rank 0 stores state to disk)
3. disconnect from upstream peer (all but rank 0)

Notes:
- this should work for entire instance or partial subtree
- timeout (e.g shutdown grace) would be on peer disconnect not on time to run rc3 (is it global or does it cascade per TBON level?  More thought needed)
- add `flux admin shutdown [OPTIONS]` command for controlled full or partial shtudown
- wait for running work to complete, or configurable timeout then fatal job exception
- support for allowing downstream peers to continue operating while disconnected TBD

Current system
- fully wires up to all ranks before executing rc1 (requires all nodes to be up)
- rc1 runs on rank 0, performs actions across ranks (requires all nodes to be up, nodes cannot join late)
- rc3 runs on rank 0, performs actions across ranks, runs under arbitrary timeout (requires all nodes to be up, doesn't wait for correct but slow operations such as big KVS flush to disk)

**Comment by garlick** [comment]
I ended up implementing (not yet submitted) a slightly modified startup sequence from what was described above.   Instead of starting the next level of brokers by accepting their TBON connections, I let the connections be made early, and implement an RPC handshake in the runlevel code called `runlevel.wait-ready`.

A broker responds to a `runlevel.wait-ready` request once it has entered run level 2 (success) or run level 3 (failure).

The rank 0 broker immediately enters runlevel 1, runs its rc script, and progresses to runlevel 2 or 3.

A broker of rank > 0 enters runlevel 1, but makes the `runlevel.wait-ready` RPC and doesn't start the rc script until it receives a response.

This achieves the goal of allowing the instance to be "up" as soon as rank 0 completes its rc1, with other ranks joining later (possibly far in the future or never).  It preserves the upstream before downstream loading order of modules.  Unlike the approach described above, wireup can complete in parallel, and I think keeping this synchronization out of the wireup code a bit easier to manage in the broker code as it is now.

As described in #2511, the intent is to have an eventlog that records nodes (execution targets) as they join the instance, and for schedulers to watch this eventlog to determine when an execution target is available for scheduling.

One problem this new scheme creates is that hwloc is currently collected in rc1 on the rank 0 broker and assumes all nodes are up.  Not only does rc1 on rank 0 now execute before other ranks are up, but nodes might be remain down.  I've open #2908 to track this problem.

**Comment by garlick** [comment]
The details of shutdown are not stable yet in my branch, but the idea is to publish a shutdown event.  Upon receipt of this event, each broker enters rc3 but delays running its rc3 script until any downstream peers disconnect.   When rc3 completes, a broker disconnects and exits.

This preserves the downstream before upstream unloading order of modules.

One problem is that we currently have a kind of "prologue" at the top of the rank 0 rc3:
```sh
flux queue stop --quiet
flux job cancelall --user=all --quiet -f --states RUN
flux queue idle --quiet
```
This won't work anymore because rc3 on rank 0 executes _last_.   I've opened #2909 for discussion of this problem.

**Comment by garlick** [comment]
Closed by #3168.
